### ISSUE TITLE
Every non-quantized model loads in float32: from_pretrained gets no dtype, so a bf16 checkpoint costs 2x

### ISSUE THREAD

**Issue by MakazhanAlpamys** [MakazhanAlpamys/Soup#339]
`SFTTrainerWrapper` builds `model_kwargs` with `trust_remote_code` and `device_map` and **no dtype**, on all three modality paths (`sft.py:860`, `:1105`, `:1214`). `from_pretrained` then defaults to `float32` regardless of what the checkpoint stores.

Verified directly:

```
checkpoint config dtype : bfloat16
Soup's call (no dtype)  : torch.float32
with dtype='auto'       : torch.bfloat16
bytes per param   soup: 4    auto: 2
```

So a bf16 checkpoint occupies **twice** the VRAM it needs to. `training_args` sets `bf16=True`, which makes the *compute* bf16 via autocast, but the master weights stay fp32 — that is invisible in the loss and very visible in `nvidia-smi`.

### Why this is not simply "add dtype='auto'"

For **full fine-tuning** fp32 master weights are the standard mixed-precision arrangement and dropping them changes optimisation, not just memory. For a **frozen base under LoRA** — which is what most Soup runs are — holding the base in fp32 is pure waste: it is never updated, and the adapter is what trains.

So the fix is probably conditional (frozen base -> checkpoint dtype; trainable base -> keep fp32 master weights), and it is a numerics decision, not a one-liner. Filing rather than changing it unilaterally.

### Where it was found
Measuring #306 at 7B. It is also the reason 8B full-FT cannot run on an 80 GB card at all: 8.03B x 16 B/param = **119.7 GB** of weights + grads + AdamW state before a single activation. At bf16 master weights the same arithmetic is 8.03B x 12 = 96.4 GB — still over, but the gap is entirely this default.

### Acceptance
- [ ] a LoRA run on a bf16 checkpoint loads the frozen base at the checkpoint's dtype
- [ ] full fine-tuning keeps whatever precision arrangement is deliberately chosen, documented
- [ ] a test asserting the loaded dtype for both cases
- [ ] the VRAM pre-flight's assumptions re-checked against whichever choice lands

Record: `benchmarks/gate-h100-validation.md` STEP 22.

**Comment by MakazhanAlpamys** [comment]
Independently corroborated while measuring #314 on the same box.

A full-fine-tuning pilot on Llama-3.1-8B OOM'd at `optimizer.step()` on an 80 GB card, and the arithmetic names this issue as the cause: fp32 load gives **32 GB weights + 32 GB grads + 16 GB 8-bit Adam ~= 80 GB**, before activations. At the checkpoint's own bf16 the same run would have had headroom.

So the 2x is not only a memory-efficiency question — it is currently the difference between a full fine-tune running and not running at 8B.

**Comment by MakazhanAlpamys** [comment]
## Measured on the H100 box: the cost on a real LoRA run

The issue established the defect and the arithmetic; what it did not have was a
number on the configuration most Soup runs actually are — LoRA, base frozen,
where fp32 master weights are pure waste. Measured now.

Llama-3.1-8B, `quantization: none`, LoRA r=16 on `q_proj`/`v_proj`, seq 512,
batch 1, 20 steps, AdamW. Two arms in one process, each freed before the next
builds. `src/` is frozen for v0.73.0, so the `dtype='auto'` arm is produced by
wrapping `from_pretrained` at runtime rather than by editing `sft.py`.

**With autocast — the path a real run takes** (`training_args` sets `bf16=True`):

| | base dtype | peak VRAM | median tok/s |
|---|---|---|---|
| as shipped (no dtype) | `torch.float32` | **48 241 MiB** | 6 989.8 |
| `dtype='auto'` | `torch.bfloat16` | **18 658 MiB** | 6 568.3 |

**28.9 GB saved, a 2.59x ratio.** Not 2x: autocast materialises bf16 copies on
top of the fp32 masters, so the shipped path pays for both.

Throughput goes the other way — the bf16 arm is **6% slower** (0.94x). Reported
because it was measured, not explained: nothing here predicts it, and it is the
opposite of the "fp32 moves twice the bytes" intuition. One operating point,
n=1, so treat it as a direction rather than a figure.

### A first version of this measurement was wrong, and is recorded rather than dropped

Run without autocast, the same two arms gave 36 769 -> 18 902 MiB (1.95x) and the
bf16 arm **5.31x faster**. That 5.31x is an artefact of the harness: with no
autocast the fp32 arm does fp32 matmuls, which is not the path Soup runs. It is a
number about autocast, not about dtype. The autocast table above supersedes it.

### Guards

- The loaded base dtype is asserted per arm (`float32` vs `bfloat16`); equal
  dtypes raise `VACUOUS` rather than reporting a delta.
- Loss curves compared across arms: **max relative difference 0.00039**. They are
  not there to judge quality — 20 steps says nothing about that — but to show the
  arms trained the same thing, so the VRAM delta is attributable to the dtype.
- Peak is `max_memory_allocated` after `reset_peak_memory_stats`, per arm, with
  the previous arm freed and `empty_cache()` called.

### What this does and does not settle

It sizes the waste; it does not choose the fix. The issue's own argument stands:
frozen base -> checkpoint dtype, trainable base -> keep fp32 master weights,
which is a numerics decision rather than a one-liner. What is now known is that
on the frozen-base path the decision is worth **28.9 GB at 8B**, and that
throughput is not a free win in that direction.

Environment: torch 2.13.0+cu130, transformers 4.57.6, H100 80GB HBM3.
Raw: `/root/results/dtype_cost_8b_autocast.json` (autocast) and
`/root/results/dtype_cost_8b.json` (the superseded no-autocast run).

**Comment by MakazhanAlpamys** [comment]
## Correction to the comment above: the throughput direction was n=1 noise

I reported "the bf16 arm is 6% slower (0.94x)" from a single run and called it a
direction. Two more repeats destroy it.

| repeat | fp32 tok/s | bf16 tok/s | ratio |
|---|---|---|---|
| 1 | 6 989.8 | 6 568.3 | 0.94 |
| 2 | 6 643.6 | 7 380.1 | **1.11** |
| 3 | 6 891.1 | 7 418.6 | **1.08** |

Two of three go the other way. The within-arm spread (6 568–7 419 = 13% inside
the bf16 arm alone) is **larger than the between-arm difference**, so at n=3 this
box cannot resolve a throughput effect in either direction. The honest statement
is that **the throughput impact is unresolved**, not that bf16 costs 6%.

Withdrawn rather than edited in place, because the first number was published.

## What n=3 did confirm, byte for byte

| repeat | peak fp32 | peak bf16 |
|---|---|---|
| 1 | 48 241 MiB | 18 658 MiB |
| 2 | 48 241 MiB | 18 658 MiB |
| 3 | 48 241 MiB | 18 658 MiB |

Identical to the byte across all three — `max_memory_allocated` is deterministic
for a fixed shape, so the repeats confirm determinism rather than estimate noise.

**The 28.9 GB / 2.59x saving stands unchanged.** It is the load-bearing number in
this issue and it is the one that reproduced exactly; the throughput figure was
the incidental one and it is the one that did not.

**Comment by blackcoderx** [comment]
Picking this up — will follow the frozen-base -> checkpoint dtype (`dtype="auto"`) / trainable-base -> explicit `torch.float32` split you outlined, thread it through all three `SFTTrainerWrapper` load sites (text/vision/audio), and re-check `hardware_fit.py`'s bytes-per-param assumption against it. Will flag the `commands/train.py` peft-classifier gap (missing `lisa_enabled`/`lora.r==0`) inline rather than fixing it here, since it's a pre-existing issue independent of this one.

**Comment by MakazhanAlpamys** [comment]
﻿Shipped by #471 (@blackcoderx), merged on a fully green 13-cell CI run.

What you get concretely: a frozen base now loads in its checkpoint'"'"'s own dtype instead of float32, so a bf16 checkpoint stops costing 2x memory; full-FT paths (`lora.r: 0`, Spectrum, LISA) load explicit fp32 master weights; and the pre-flight classifies full-FT through the same predicate the trainer uses. Try: `soup train --config soup.yaml` with a bf16 base and watch the reported load dtype / memory drop.

The twelve non-SFT trainers were already covered by #492; this closes the SFT-side half plus the pre-flight drift.
